src/QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp

```diff
@@ -79,66 +79,60 @@ class DiracMatrixComputeCUDA : public Resource
    *  \param[in,out]  a_mats      dual A matrices, they will be transposed on the device side as a side effect.
    *  \param[out]     inv_a_mats  dual invM matrices
    *  \param[in]      n           matrices rank.								
-   *  \param[in]      lda         leading dimension of each matrix					
    *  \param[out]     log_values  log determinant value for each matrix, batch_size = log_values.size()
    *
-   *  This is the prefered interface for calling computeInvertAndLog when the VALUE_FP and TMAT are equal.
    *  On Volta so far little seems to be achieved by having the mats continuous.
    */
-  template<typename TMAT>
   inline void mw_computeInvertAndLog(CUDALinearAlgebraHandles& cuda_handles,
-                                     RefVector<DualMatrix<TMAT>>& a_mats,
-                                     RefVector<DualMatrix<TMAT>>& inv_a_mats,
+                                     RefVector<DualMatrix<VALUE_FP>>& a_mats,
+                                     RefVector<DualMatrix<VALUE_FP>>& inv_a_mats,
                                      const int n,
-                                     const int lda,
                                      DualVector<LogValue>& log_values)
   {
-    int nw = a_mats.size();
+    const int nw = a_mats.size();
     assert(a_mats.size() == inv_a_mats.size());
 
-    psiM_ptrs_.resize(sizeof(TMAT*) * nw);
-    invM_ptrs_.resize(sizeof(TMAT*) * nw);
-    //temp_mat_.resize(n,lda);
-    int ldinv = inv_a_mats[0].get().cols();
-    Vector<const VALUE_FP*> M_ptr_buffer(reinterpret_cast<const TMAT**>(psiM_ptrs_.data()), nw);
-    Vector<const VALUE_FP*> invM_ptr_buffer(reinterpret_cast<const TMAT**>(invM_ptrs_.data()), nw);
+    psiM_ptrs_.resize(nw);
+    invM_ptrs_.resize(nw);
+    const int lda           = a_mats[0].get().cols();
+    const int ldinv         = inv_a_mats[0].get().cols();
     cudaStream_t hstream    = cuda_handles.hstream;
     cublasHandle_t h_cublas = cuda_handles.h_cublas;
+    psiM_fp_.resize(n * ldinv * nw);
```

Comment:
hmm this was only used if VALUE != VALUE_FP since then we can just use the inplace a_mats.  Is this the issue?  Is a_mat assumed to not be updated since the CPU lapack transpose in the the inv matrix and eventually does the tri in place?  So a_mat is assumed to be persistent state for the Determinant even though it is named _temp and appears to be compute from Phi->evaluate_notranspose?

Reply:
batched LU needs the same padding for matrix_in and matrix_inv. However, a_mats don't have padding but inv_a_mats have. So use psiM_fp_ to store a_mats_transpose with padding. The additional benefit is a_mats are not destroyed.

Reply:
I think it is better to introduce separate LDA for in and out matrix in cuBLAS_LU to make it more flexible. However, I also prefer to keep using psiM_fp_ because I'm not comfortable with input psiM being destroyed.

Reply:
I agree.

Reply:
I pushed some notes. If we decide to use psiM_fp_ then we can do the transpose on the host and fuse transfer like the mixed precision code path. In this way, matrix by matrix cuBLAS::geam can be dropped as well.

Reply:
I initially did the transposition on CPU but benchmarking showed a large performance loss so I moved it. I wanted to show the the batched inversion on GPU was worth doing at least some of the time and geam was required for that. The real problem is the storage layout should not be row major all this transposing would be unecessary.  So the update is performant one way and the lapack another.  I think the update code could be written as easily with columns instead of rows though.